src/index.js

```diff
@@ -387,14 +403,209 @@ function fixTouchableMixin(t, path) {
+  if (path.node.name.name === 'ref') {
+    // ignore if the ref value is already applyFSPropertiesWithRef
+    if (
+      t.isJSXExpressionContainer(path.node.value) &&
+      !(
+        t.isCallExpression(path.node.value.expression) &&
+        path.node.value.expression.callee.name === applyFSPropertiesWithRef
+      )
+    ) {
+      // only process refs on base components and applyFSPropertiesWithRef must be imported
+      if (
+        !elements.includes(path.parent.name.name) ||
+        !path.scope.hasBinding(applyFSPropertiesWithRef)
+      )
+        return;
+
+      const originalRef = path.node.value.expression;
+
+      path.replaceWith(
+        t.jsxAttribute(
+          t.jsxIdentifier('ref'),
+          t.JSXExpressionContainer(
+            t.CallExpression(t.identifier(applyFSPropertiesWithRef), [originalRef]),
+          ),
+        ),
+      );
+    }
+  }
+}
+
 /* eslint-disable complexity */
 export default function ({ types: t }) {
   return {
+    name: 'fullstory-react-native',
     visitor: {
       /* Looks like we don't currently need to add this to the interface declaration
       InterfaceDeclaration(path, state) {
         fixUIManagerJSInterface(path);
       }, */
+
+      Program: {
+        enter(path, { file, opts }) {
+          if (opts.isNewArchitectureEnabled) {
+            const hasReactNativeOrReact = path.node.body.some(x => {
+              return (
+                t.isImportDeclaration(x) &&
+                (x.source.value === 'react-native' || x.source.value === 'react')
+              );
+            });
+            // Do nothing if applyFSPropertiesWithRef is already declared
+            // React Native needs to be in scope
+            if (path.scope.hasBinding(applyFSPropertiesWithRef) || !hasReactNativeOrReact) {
+              return;
+            }
+
+            // create applyFSPropertiesWithRef import statement
+            const applyFSImportStatement = t.importDeclaration(
+              [
+                t.importSpecifier(
+                  t.identifier(applyFSPropertiesWithRef),
+                  t.identifier(applyFSPropertiesWithRef),
+                ),
+              ],
+              t.stringLiteral('@fullstory/react-native'),
+            );
+
+            // add import statement to top of file
+            const [newPath] = path.unshiftContainer('body', applyFSImportStatement);
+
+            // register import in scope
+            newPath.get('specifiers').forEach(specifier => {
+              path.scope.registerBinding('module', specifier);
+            });
+
+            // save import node in state
+            file.set('ourPath', newPath);
+
+            // Add a placeholder variable bount to applyFSPropertiesWithRef to prevent
+            // unused variable deletion from other plugins (@babel/plugin-transform-typescript)
+            const placeholder = t.variableDeclaration('let', [
+              t.variableDeclarator(
+                t.identifier('handle'),
```

Comment:
is there a chance this could clash or override with any other `handle` declaration?

Reply:
Good catch! Renamed the variable to something more unique